### DISCUSSION TITLE
[SAMR34] RtcMs2Tick conversion not precise

### DISCUSSION THREAD

**Discussion by KaDw** [Lora-net/LoRaMac-node#1045] [Q&A] [answered]
Hi, 

In SAMR34 port function RtcMs2Tick() is not taking into acount that the RTC clock is in fact 1,024kHz not 1kHz. If we set software timer to 1000ms we loose about 24ms (If I got it right)
What is more, there are two functions RtcMs2Tick and RtcTick2Ms. RtcTick2Ms takes that into account, RtcMs2Tick does not. 

```
uint32_t RtcMs2Tick( TimerTime_t milliseconds )
{
    return ( uint32_t )( milliseconds );
}

TimerTime_t RtcTick2Ms( uint32_t tick )
{
    uint32_t seconds = tick >> 10; // seconds = tick / 1024

    tick = tick & 0x3FF; // mask to 1024
    return ( ( seconds * 1000 ) + ( ( tick * 1000 ) >> 10 ) );
}
```

I changed it to (we lose some precision anyway):

```
uint32_t RtcMs2Tick( TimerTime_t milliseconds )
{
    uint32_t tick = (milliseconds << 10) / 1000;

    return ( uint32_t )( tick );
}
```

Now when I measure it with scope it is closer to value set by TimerSetValue().

Could you explain why the time is not converted? The stack must open the RX1 and RX2 windows precisely and the timer will cause about 24 and 48ms offset respectively!

**Comment by mluis1** [comment, marked as answer]
Thanks for reporting this issue.

I would propose the following

```c
uint32_t RtcMs2Tick( uint32_t milliseconds )
{
    return ( uint32_t )( ( ( ( uint64_t )milliseconds ) << 10 ) / 1000 );
}
```

By applying the shift left by ten the result may not fit in a `uint32_t` type.

Could you please create a pull request?

**Reply by KaDw** [reply]
I traced all `RtcMs2Tick` calls and stumbled upon this function:

```
TimerTime_t TimerGetElapsedTime( TimerTime_t past )
{
    if ( past == 0 )
    {
        return 0;
    }
    uint32_t nowInTicks = RtcGetTimerValue( );
    uint32_t pastInTicks = RtcMs2Tick( past );

    // Intentional wrap around. Works Ok if tick duration below 1ms
    return RtcTick2Ms( nowInTicks - pastInTicks );
}
```

I'm just wondering what will happen if the `past` variable will have value will be above 4 194 304 000 (1000/1024*UINT32_MAX). If we pass this value to `RtcMs2Tick` it will overflow. Lets assume that `nowInTicks` is 4 194 306 000 and `pastInTicks` is 4 194 304 001. The `TimerGetElapsedTime` will return 4 194 306 000 - 1 = 4 194 305 999 instead of 1999

**Reply by KaDw** [reply]
Ok, I plugged in debugger and set RTC COUNT to value close to overflow, I turns out that `nowInTicks` is smaller than `pastInTicks`, I don't understand why is that but in the scenario described [above ](https://github.com/Lora-net/LoRaMac-node/discussions/1045#discussioncomment-380250)`pastInTicks` is always grater than `nowInTicks` :confused:

**Reply by Tom1278** [reply]
Hi, my RtcMs2Tick was also not correct and resulted in timing issues. The 5 second RX1 delay for example was off by as much as 200ms.
My new function looks like this and it's accurate to about 2ms:
uint32_t RtcMs2Tick( uint32_t milliseconds )
{
	uint32_t ret;
    ret = ( uint32_t )( ( ( ( uint64_t )milliseconds ) * CONV_DENOM ) / CONV_NUMER );
    ret -= (milliseconds / 97);
    return ret;
}
